compiler/rustc_middle/src/lint.rs

```diff
@@ -173,7 +173,7 @@ impl TyCtxt<'_> {
 /// This struct represents a lint expectation and holds all required information
 /// to emit the `unfulfilled_lint_expectations` lint if it is unfulfilled after
 /// the `LateLintPass` has completed.
-#[derive(Clone, Debug, HashStable)]
+#[derive(Clone, Debug, Encodable, Decodable, HashStable)]
 pub struct LintExpectation {
```

Comment:
`CodegenResults::serialize_rlink` doesn't serialize the source map. While it this is fully correct and shouldn't be changed, this means that when a `LintExpectation` inside `CodegenResults` gets deserialized in a new compiler session, the `Span` inside of it will be garbage and similarly for the `Span` in `LintLevelSource`. Would it be possible to avoid serializing the spans? Or if not, replace them with `DUMMY_SP`? I'm not sure how well the linting infrastructure handles garbage spans.

Reply:
> I'm not sure how well the linting infrastructure handles garbage spans.

Poorly. If the spans are empty or hi < lo, I think that triggers debug assertions.

Reply:
The raw lo and hi values are preserved across the serialization roundtrip (so the lo <= hi invariant is preserved (empty spans are valid and rendered as a span from lo to lo + 1)), however they refer to non-existent parts of the source map.

Reply:
Hm, if those become OOB, then I think they also trigger debug assertions (or actual assertions) if there's any lint that unwraps when trying to read a SourceMap snippet.

Reply:
this indeed crashes currently:
```
=== STDERR ===
thread 'rustc' panicked at compiler/rustc_span/src/source_map.rs:1056:40:
index out of bounds: the len is 0 but the index is 18446744073709551615
```

i thought i could solve this by deserializing to a dummy span but that doesn't seem to have worked for some reason? it still gives the same error.
```
  INFO rustc_codegen_ssa::back::link Deny Node { name: "linker_messages", span: no-location (#0), reason: None }
```

i *think* the right fix here is not to use a source span map at all during `-Z link-only`, to have the emitter return None in `Emitter::source_map`. but i am not sure exactly how to set that up.